### DISCUSSION TITLE
同时使用两个IP地址分别进行单向传输代理数据

### DISCUSSION THREAD

**Discussion by OnlyCharacters** [XTLS/Xray-core#1332] [Ideas]
我们平时使用代理都是建立在同一个网络层协议栈上的，要么是 IPv4，要么是 IPv6，往返都是同一对 IP 地址。

设想一下，如果客户端和服务器都有两个 IP 地址，那么为什么不尝试使用两对 IP 地址进行通信？一对用于客户端发送请求，一对用于服务器返回响应数据，这样情况下是不是中间人/审查者就很难从单个连接来判断客户端是否使用代理了？

我写了一个很简单的 demo 来实现上述的通信模型，[demo地址](https://github.com/OnlyCharacters/ParallelSocket)，不知道这种方式是否对抗审查有作用，希望各位来讨论一下。

**Comment by yuhan6665** [comment]
我觉得你的想法很有意思 如果能实现 有很多新的可能性 推荐你看一下相关的英文研究
https://github.com/net4people/bbs/issues/74

**Comment by cross-hello** [comment]
So interesting

----------------------------------------

*From: *OnlyCharacters ***@***.***>
*To: *XTLS/Xray-core ***@***.***>
*CC: *Subscribed ***@***.***>
*Date: *Nov 14, 2022 03:03:51
*Subject: *[XTLS/Xray-core] 同时使用两个IP地址分别进行单向传输代理数据 (Discussion #1332)

**Comment by zxbiao** [comment]
如果VPS拥有2个或多个IP地址时，可以搞这个，一个出国多个回国，越多IP地址就越难判别关系。
能防止GFW对同一IP地址出入的流量做分析，大大加大了GFW的审查难度。
但从另一方面来想
**长时间对某个IP发出数据而不返回数据**
**而某个IP也长时间返回数据而不发出数据**也算是一个特征
可能GFW要分析很长时间才能确定到相应的关系。

**Comment by kulongwangzhi85** [comment]
如果你熟悉linux的防火墙，可以玩出花来。
v2ray开多个inbound，防火墙标记tcp-syn，如果有两条出口，就1/2切换一次线路。如果有3条出口，就1/3切换一次线路

目前我使用ipv4走grpc，ipv6走quic，v2ray路由内指定tcp走grpc，udp走quic

**Reply by yuanweize** [reply]
有教程吗?想用cfwarp➕本地双出口

**Comment by testcaoy7** [comment]
Usually asymmetric routing is filtered by stateful firewall.
https://harrymaq.medium.com/symmetric-and-asymmetric-routing-with-example-78b8047b5f29

The demo uses ipv6 for returning packets, maybe because most ipv6 is stateless protocol?
If you do NAT66, the demo may not work.

**Comment by nlifew** [comment]
和我现在的思路差不多哈哈哈，不过我的想法是一个主信道若干个数据信道，需要建立新链接时服务端分配两个数据信道，一个上行一个下行，这两个完全可以使用完全不同的协议，比如上行使用 udp 下行使用 tcp 这种
